### PR TITLE
Add metrics to the http json service

### PR DESCRIPTION
**Pull request by realvictorprm**
### Pull Request Checklist

Just adds metrics to the http json service.

Also moves out some metrics related stuff to the Metrics project so the http json service does not depend on unrelated services.

changelog_begin

[HTTP/JSON API]
- metrics reporting can now be chosen via the command line option --metrics-reporter (currently hidden), valid options are console, csv://, graphite:// and prometheus://
- metrics reporting interval can also now be chosen via a command line option --metrics-reporting-interval (currently hidden)

changelog_end

Minimised version of try 2.

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala [hunk 1]
```diff
@@ -682,5 +682,12 @@ final class Metrics(val registry: MetricRegistry) {
         val prune: Timer = registry.timer(Prefix :+ "prune")
       }
     }
+
+    object http_json_api {
```
Comment (miklos-da) [inline ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala:686]: Please don't use underscore in class or object names.
```suggestion
    object httpJsonApi {
```

File: ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala [hunk 2]
```diff
@@ -309,6 +300,11 @@ class CommonCliBase(name: LedgerName) {
           s"By default compression is off, this switch enables it. This has only effect for append-only ingestion." // TODO append-only: fix description
         )
 
+      com.daml.cliopts.Metrics.metricsReporterParse(this)(
+        (f, c) => c.copy(metricsReporter = f(c.metricsReporter)),
```
Comment (miklos-da) [inline ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala:304]: To be consistent with the style already established, please don't use abbreviations:
```suggestion
        (setter, config) => config.copy(metricsReporter = setter(config.metricsReporter)),
```

**Review by miklos-da — CHANGES_REQUESTED** [review]
Thanks for greatly simplifying the original PR #9888! Just requesting minor changes, otherwise LGTM.

File: ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala (hunk 1 above)
Comment (realvictorprm) [reply ledger/metrics/src/main/scala/com/daml/metrics/Metrics.scala]: Adressed with 169de2eb1b310000636dac82e4e4e72da476044c

File: ledger/metrics/src/main/scala/com/daml/metrics/MetricsReporter.scala [hunk 3]
```diff
@@ -89,17 +89,17 @@ object MetricsReporter {
...
+  val cliHint: String =
+    """Must be one of "console", "csv:///PATH", "graphite://HOST[:PORT][/METRIC_PREFIX]", or "prometheus://HOST[:PORT]"."""
+
   def parseUri(value: String): URI =
     try {
       new URI(value)
     } catch {
       case NonFatal(exception) =>
-        throw new InvalidConfigException(cliHint + " " + exception.getMessage)
+        throw new RuntimeException(cliHint, exception)
```
Comment (cocreature) [inline ledger/metrics/src/main/scala/com/daml/metrics/MetricsReporter.scala:100]: Seems like a reasonable change, nobody should be catching this anyway.

**Review by cocreature — APPROVED** [review]
Nice, thank you!

**Review by miklos-da — APPROVED** [review]

File: ledger/sandbox-common/src/test/lib/scala/platform/sandbox/cli/CommonCliSpecBase.scala [hunk 4]
```diff
@@ -268,8 +268,8 @@ abstract class CommonCliSpecBase(
 
     "parse the metrics reporting interval when given" in {
       checkOption(
-        Array("--metrics-reporting-interval", "PT1M30S"),
-        _.copy(metricsReportingInterval = Duration.ofSeconds(90)),
+        Array("--metrics-reporting-interval", "1.5m"),
```
Comment (realvictorprm) [inline ledger/sandbox-common/src/test/lib/scala/platform/sandbox/cli/CommonCliSpecBase.scala:271]: @cocreature Is this cli option change OK? This is due to the change from Java Duration to scala duration

Comment (cocreature) [reply ledger/sandbox-common/src/test/lib/scala/platform/sandbox/cli/CommonCliSpecBase.scala:271]: This seems like a breaking change. Can we continue accepting a Java duration here? Or alternatively accept both preferring a Java duration. We can still convert to a Scala duration at the very beginning if that helps.

Comment (cocreature) [reply ledger/sandbox-common/src/test/lib/scala/platform/sandbox/cli/CommonCliSpecBase.scala:271]: You should be able to use a custom reader that tries one first and if that fails tries the other.

Comment (realvictorprm) [reply ledger/sandbox-common/src/test/lib/scala/platform/sandbox/cli/CommonCliSpecBase.scala:271]: Yeah just wrote one, I will make another PR as I forgot to remove the label here.

**Pull request merged**